### PR TITLE
Working Code

### PR DESCRIPTION
**Pull request by liam-middlebrook**
I was unable to get the HTTP `Referer` header working properly which is why I use a querystring to pass around the redirect.

Not 200-500 lines, if that matters don't review it.

Working Demo: https://go-auth-test.csh.rit.edu
Source for Demo: https://github.com/liam-middlebrook/golang-openshift-test

### CODE REVIEW

File: csh_auth.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    log "github.com/sirupsen/logrus"
+    jwt "github.com/dgrijalva/jwt-go"
+    oidc "github.com/coreos/go-oidc"
+    "golang.org/x/net/context"
+    "golang.org/x/oauth2"
+    "github.com/gin-gonic/gin"
+)
+
+
+var clientID = os.Getenv("csh_auth_client_id")
```
Comment (rossdylan) [inline csh_auth.go]: gin supports the idea of middleware components which can be instantiated by the user. I'd recommend looking through that. You can create a middlewhere constructor that pulls all this data into a config that way all the globals are removed.

File: csh_auth.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+var clientSecret = os.Getenv("csh_auth_client_secret")
+var secret = os.Getenv("csh_auth_jwt_secret")
+var state = os.Getenv("csh_auth_state")
+var server_host = os.Getenv("csh_auth_server_host")
+var redirect_uri = os.Getenv("csh_auth_redirect_uri")
+
+var AuthKey = "cshauth"
```
Comment (rossdylan) [inline csh_auth.go]: const for constant values

File: csh_auth.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+var provider_uri = "https://sso.csh.rit.edu/realms/csh"
+
+var authenticate_uri = "/authenticate"
+// globals eww
+var config oauth2.Config // this guy changes a bit, weird
+var ctx context.Context
```
Comment (rossdylan) [inline csh_auth.go]: contexts shouldn't be long lived like this. They are usually scoped to the lifetime of a single request or a background task.

File: csh_auth.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
+package csh_auth
```
Comment (rossdylan) [inline csh_auth.go:1]: nit: run `go fmt` to properly order imports, and fix alignments on struct definitions.

File: csh_auth.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        cookie, err := c.Cookie("Auth")
+        if err != nil || cookie == "" {
+            log.Info("cookie not found")
+            c.Redirect(http.StatusFound, authenticate_uri + "?referer=" + c.Request.URL.String())
+            return
+        }
+
+        token, err := jwt.ParseWithClaims(cookie, &CSHClaims{}, func(token *jwt.Token) (interface{}, error){
+            if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+                return nil, fmt.Errorf("unexpected signing method")
```
Comment (rossdylan) [inline csh_auth.go]: static error message, use errors.New

File: csh_auth.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        }
+
+        if claims, ok := token.Claims.(*CSHClaims); ok && token.Valid {
+            // add in user info data
+            c.Set(AuthKey, *claims)
+            // call the wrapped func
+            page(c)
+        } else {
+            log.Error("claim parsing failure")
+            return
```
Comment (rossdylan) [inline csh_auth.go]: nit: unnecessary return

File: csh_auth.go (hunk 4 above)
Comment (rossdylan) [reply csh_auth.go:1]: I don't remember which editor you prefer, but for vim `vim-go` and the `go-metalinter` are great tools for catching all the misc style issues, and tiny errors.

**Review by rossdylan — COMMENTED** [review]
Base code isn't too bad. However I believe you can improve the structure greatly by looking into gin's middleware system. It should allow for a more flexible non-globals based system

**Comment by saucetray** [comment]
Please merge this to master so I can use go get

**Pull request merged**